File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/exceptions/LoggerUncaughtExceptionHandler.java
package com.lasthopesoftware.bluewater.shared.exceptions;

import com.lasthopesoftware.resources.executors.CachedSingleThreadExecutor;
import com.namehillsoftware.handoff.rejections.UnhandledRejectionsReceiver;
import com.namehillsoftware.lazyj.CreateAndHold;
import com.namehillsoftware.lazyj.Lazy;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.Thread.UncaughtExceptionHandler;
import java.util.concurrent.Executor;

public class LoggerUncaughtExceptionHandler implements UncaughtExceptionHandler, UnhandledRejectionsReceiver {

	private static final CreateAndHold<Executor> errorExecutor = new Lazy<>(CachedSingleThreadExecutor::new);

	private static final Logger logger = LoggerFactory.getLogger(LoggerUncaughtExceptionHandler.class);

	@Override
	public void uncaughtException(Thread thread, Throwable ex) {
		getErrorExecutor().execute(() -> logger.error("Uncaught Exception", ex));
	}

	@Override
	public void newUnhandledRejection(Throwable rejection) {
		getErrorExecutor().execute(() -> logger.warn("An asynchronous exception has not yet been handled", rejection));
	}

	public static Executor getErrorExecutor() {
		return errorExecutor.getObject();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/exceptions/LoggerUncaughtExceptionHandler.kt
package com.lasthopesoftware.bluewater.shared.exceptions

import com.lasthopesoftware.resources.executors.ThreadPools
import com.namehillsoftware.handoff.rejections.UnhandledRejectionsReceiver
import org.slf4j.LoggerFactory
import java.net.SocketException
import java.util.Locale
import java.util.concurrent.CancellationException

private const val UnhandledCancellationException = "A CancellationException was unhandled"
private const val UnhandledSocketClosedException = ""

object LoggerUncaughtExceptionHandler : Thread.UncaughtExceptionHandler, UnhandledRejectionsReceiver {

	private val logger by lazy {
		LoggerFactory.getLogger(javaClass).apply {
			info("Uncaught exceptions logging to custom uncaught exception handler.")
		}
	}

	override fun uncaughtException(thread: Thread, ex: Throwable) {
		if (reportException(ex)) {
			ThreadPools.exceptionsLogger.execute { logger.error("Uncaught Exception", ex) }
		}
	}

	override fun newUnhandledRejection(rejection: Throwable) {
		if (reportException(rejection)) {
			ThreadPools.exceptionsLogger.execute {
				logger.warn("An asynchronous exception has not yet been handled", rejection)
			}
		}
	}

	private fun reportException(ex: Throwable): Boolean = when (ex) {
			is CancellationException -> {
				logger.debug(UnhandledCancellationException, ex)
				false
			}
			is SocketException -> {
				ex.message
					?.takeIf { it.lowercase(Locale.getDefault())  == "socket closed" }
					?.let {
						logger.debug(UnhandledSocketClosedException, ex)
						false
					}
					?: true
			}
			else -> true
		}
}
